File: src/prototype/web/tests/posture_break.test.ts
// Tests for posture_break updates — docs/design/state_machines_v1.md §3.

import { describe, expect, it } from "vitest";
import {
  DEFAULT_POSTURE_CONFIG,
  breakBucket,
  breakMagnitude,
  gripPullVector,
  updatePostureBreak,
  type PostureBreakInputs,
} from "../src/state/posture_break.js";

function inputs(over: Partial<PostureBreakInputs> = {}): PostureBreakInputs {
  return {
    dtMs: 16.67,
    attackerHip: { hip_angle_target: 0, hip_push: 0, hip_lateral: 0 },
    gripPulls: [],
    defenderRecovery: { x: 0, y: 0 },
    ...over,
  };
}

describe("posture_break decay (§3.3 bullet 1)", () => {
  it("zero input + zero state stays zero", () => {
    const out = updatePostureBreak({ x: 0, y: 0 }, inputs());
    expect(out.x).toBe(0);
    expect(out.y).toBe(0);
  });

  it("800ms time constant: after one τ with no input, magnitude falls to ~1/e", () => {
    const start = { x: 0.8, y: 0 };
    // Run ~800ms in 16.67ms slices.
    let v = start;
    const step = 16.67;
    for (let t = 0; t < DEFAULT_POSTURE_CONFIG.decayTauMs; t += step) {
      v = updatePostureBreak(v, inputs({ dtMs: step }));
    }
    // e^-1 ≈ 0.3679; allow small integration slack.
    expect(v.x).toBeGreaterThan(0.8 * 0.35);
    expect(v.x).toBeLessThan(0.8 * 0.40);
  });
});

describe("posture_break attacker hip contribution (§3.3 bullet 2)", () => {
  it("forward hip push over time accumulates sagittal break", () => {
    let v = { x: 0, y: 0 };
    // 1 second of hip_push = 1.0 with zero decay fight.
    for (let t = 0; t < 1000; t += 16.67) {
      v = updatePostureBreak(v, inputs({ attackerHip: { hip_angle_target: 0, hip_push: 1, hip_lateral: 0 } }));
    }
    // Equilibrium for a constant driving input u is u·τ·(K/τ) → K scaled
    // by the integration. We only assert "meaningfully positive, bounded
    // by unit clamp" because exact values are coefficient-dependent.
    expect(v.y).toBeGreaterThan(0.3);
    expect(v.y).toBeLessThanOrEqual(1);
    expect(v.x).toBe(0);
  });

  it("lateral hip drives sign of the lateral axis accordingly", () => {
    let v = { x: 0, y: 0 };
    for (let t = 0; t < 500; t += 16.67) {
      v = updatePostureBreak(v, inputs({ attackerHip: { hip_angle_target: 0, hip_push: 0, hip_lateral: -1 } }));
    }
    expect(v.x).toBeLessThan(-0.1);
  });
});

describe("posture_break grip pulls (§3.3 bullet 3)", () => {
  it("a GRIPPED sleeve adds a forward + side-directed break component", () => {
    const pull = gripPullVector("SLEEVE_R", 1);
    expect(pull.x).toBeGreaterThan(0);
    expect(pull.y).toBeGreaterThan(0);

    let v = { x: 0, y: 0 };
    for (let t = 0; t < 500; t += 16.67) {
      v = updatePostureBreak(v, inputs({ gripPulls: [pull] }));
    }
    expect(v.x).toBeGreaterThan(0);
    expect(v.y).toBeGreaterThan(0);
  });

  it("zero-strength grip contributes nothing", () => {
    const pull = gripPullVector("SLEEVE_R", 0);
    expect(pull.x).toBe(0);
    expect(pull.y).toBe(0);
  });
});

describe("posture_break defender recovery (§3.3 bullet 4)", () => {
  it("recovery input opposes existing break", () => {
    // Pre-existing forward break; recovery vector pointing forward (same
    // direction as the break) should pull it back toward origin.
    let v = { x: 0, y: 0.5 };
    for (let t = 0; t < 500; t += 16.67) {
      v = updatePostureBreak(v, inputs({ defenderRecovery: { x: 0, y: 1 } }));
    }
    // Recovery dominates decay and actually inverts the break sign here
    // because we apply -k_recovery·dt per frame with k_recovery = 1.2.
    expect(v.y).toBeLessThan(0.5);
  });
});

describe("posture_break magnitude clamp", () => {
  it("clamps to the unit disc even under sustained large input", () => {
    let v = { x: 0, y: 0 };
    for (let t = 0; t < 5000; t += 16.67) {
      v = updatePostureBreak(v, inputs({
        attackerHip: { hip_angle_target: 0, hip_push: 1, hip_lateral: 1 },
        gripPulls: [gripPullVector("COLLAR_L", 1), gripPullVector("SLEEVE_R", 1)],
      }));
    }
    expect(breakMagnitude(v)).toBeLessThanOrEqual(DEFAULT_POSTURE_CONFIG.maxMagnitude + 1e-9);
  });
});

describe("paper-proto quantization (§3.4)", () => {
  it("mags below 0.1 → bucket 0", () => {
    expect(breakBucket({ x: 0.05, y: 0 })).toBe(0);
  });
  it("mag 0.2 → bucket 1", () => {
    expect(breakBucket({ x: 0.2, y: 0 })).toBe(1);
  });
  it("mag 0.4 → bucket 2", () => {
    expect(breakBucket({ x: 0, y: 0.4 })).toBe(2);
  });
  it("mag 0.6 → bucket 3", () => {
    expect(breakBucket({ x: 0.6, y: 0 })).toBe(3);
  });
  it("mag 0.8 → bucket 4", () => {
    expect(breakBucket({ x: 0.8, y: 0 })).toBe(4);
  });
});

File: src/prototype/unity/Assets/BJJSimulator/Tests/EditMode/PostureBreakTest.cs
// NUnit EditMode mirror of src/prototype/web/tests/unit/posture_break.test.ts.
// Each [Test] corresponds to one it(...) case from the Stage 1 Vitest suite.
//
// Run from Unity Editor: Window → General → Test Runner → EditMode.

using NUnit.Framework;

namespace BJJSimulator.Tests
{
    [TestFixture]
    public class PostureBreakTest
    {
        static PostureBreakInputs Inputs(
            float dtMs          = 16.67f,
            float hipLateral    = 0f,
            float hipPush       = 0f,
            Vec2[]  gripPulls   = null,
            Vec2  defRec        = default) =>
            new PostureBreakInputs
            {
                DtMs             = dtMs,
                AttackerHip      = new HipIntent { HipAngleTarget = 0f, HipPush = hipPush, HipLateral = hipLateral },
                GripPulls        = gripPulls ?? System.Array.Empty<Vec2>(),
                DefenderRecovery = defRec.X == 0f && defRec.Y == 0f ? Vec2.Zero : defRec,
            };

        // -------------------------------------------------------------------------
        // Decay (§3.3 bullet 1)
        // -------------------------------------------------------------------------

        [Test]
        public void ZeroInput_ZeroState_RemainsZero()
        {
            var out_ = PostureBreakOps.Update(Vec2.Zero, Inputs());
            Assert.AreEqual(0f, out_.X, 1e-6f);
            Assert.AreEqual(0f, out_.Y, 1e-6f);
        }

        [Test]
        public void Decay_800ms_TimeConstant_FallsToOneOverE()
        {
            var start = new Vec2(0.8f, 0f);
            var v     = start;
            float tau  = PostureBreakConfig.Default.DecayTauMs;
            float step = 16.67f;
            for (float t = 0f; t < tau; t += step)
                v = PostureBreakOps.Update(v, Inputs(dtMs: step));
            // After one τ magnitude should be ~0.8 × e^-1 ≈ 0.294.
            Assert.Greater(v.X, 0.8f * 0.35f);
            Assert.Less(v.X,    0.8f * 0.40f);
        }

        // -------------------------------------------------------------------------
        // Attacker hip contribution (§3.3 bullet 2)
        // -------------------------------------------------------------------------

        [Test]
        public void ForwardHipPush_AccumulatesSagittalBreak()
        {
            var v = Vec2.Zero;
            for (float t = 0f; t < 1000f; t += 16.67f)
                v = PostureBreakOps.Update(v, Inputs(hipPush: 1f));
            Assert.Greater(v.Y, 0.3f);
            Assert.LessOrEqual(v.Y, 1f);
            Assert.AreEqual(0f, v.X, 1e-4f);
        }

        [Test]
        public void LateralHip_DrivesSign_OfLateralAxis()
        {
            var v = Vec2.Zero;
            for (float t = 0f; t < 500f; t += 16.67f)
                v = PostureBreakOps.Update(v, Inputs(hipLateral: -1f));
            Assert.Less(v.X, -0.1f);
        }

        // -------------------------------------------------------------------------
        // Grip pulls (§3.3 bullet 3)
        // -------------------------------------------------------------------------

        [Test]
        public void SleeveR_GrippedPull_AddsBothForwardAndSideBreak()
        {
            var pull = PostureBreakOps.GripPullVector(GripZone.SleeveR, 1f);
            Assert.Greater(pull.X, 0f);
            Assert.Greater(pull.Y, 0f);

            var v = Vec2.Zero;
            for (float t = 0f; t < 500f; t += 16.67f)
                v = PostureBreakOps.Update(v, Inputs(gripPulls: new[] { pull }));
            Assert.Greater(v.X, 0f);
            Assert.Greater(v.Y, 0f);
        }

        [Test]
        public void ZeroStrengthGrip_ContributesNothing()
        {
            var pull = PostureBreakOps.GripPullVector(GripZone.SleeveR, 0f);
            Assert.AreEqual(0f, pull.X);
            Assert.AreEqual(0f, pull.Y);
        }

        // -------------------------------------------------------------------------
        // Defender recovery (§3.3 bullet 4)
        // -------------------------------------------------------------------------

        [Test]
        public void Recovery_OpposesExistingBreak()
        {
            var v = new Vec2(0f, 0.5f);
            for (float t = 0f; t < 500f; t += 16.67f)
                v = PostureBreakOps.Update(v, Inputs(defRec: new Vec2(0f, 1f)));
            Assert.Less(v.Y, 0.5f);
        }

        // -------------------------------------------------------------------------
        // Magnitude clamp
        // -------------------------------------------------------------------------

        [Test]
        public void SustainedLargeInput_ClampsToUnitDisc()
        {
            var v = Vec2.Zero;
            var pulls = new[]
            {
                PostureBreakOps.GripPullVector(GripZone.CollarL, 1f),
                PostureBreakOps.GripPullVector(GripZone.SleeveR, 1f),
            };
            for (float t = 0f; t < 5000f; t += 16.67f)
                v = PostureBreakOps.Update(v, Inputs(hipPush: 1f, hipLateral: 1f, gripPulls: pulls));
            Assert.LessOrEqual(PostureBreakOps.BreakMagnitude(v),
                PostureBreakConfig.Default.MaxMagnitude + 1e-5f);
        }

        // -------------------------------------------------------------------------
        // Paper-proto buckets (§3.4)
        // -------------------------------------------------------------------------

        [Test] public void Bucket_BelowPoint1_IsZero()     => Assert.AreEqual(0, PostureBreakOps.BreakBucket(new Vec2(0.05f, 0f)));
        [Test] public void Bucket_Point2_IsOne()           => Assert.AreEqual(1, PostureBreakOps.BreakBucket(new Vec2(0.2f, 0f)));
        [Test] public void Bucket_Point4_IsTwo()           => Assert.AreEqual(2, PostureBreakOps.BreakBucket(new Vec2(0f, 0.4f)));
        [Test] public void Bucket_Point6_IsThree()         => Assert.AreEqual(3, PostureBreakOps.BreakBucket(new Vec2(0.6f, 0f)));
        [Test] public void Bucket_Point8_IsFour()          => Assert.AreEqual(4, PostureBreakOps.BreakBucket(new Vec2(0.8f, 0f)));
    }
}
